debug: coredump: Adds retained_memory as a backend

ldenefle opened the pull request:
This PR adds the possibility to use retained memory to store coredumps. 

It's very useful if the device does not embed a flash or if more complex processing is needed on the coredump.

It however has a couple drawbacks:
- Retained memory might be subject to power losses
- Locking mechanism on the retained memory device has to be turned off has mutexes don't work in ISR context.
- Can only store one coredump at a time, it's the responsibility of the app to read it and do something with it between different panics.

ldenefle committed:
drivers: make retained mem locking optional

Make the retained memory drivers locking mechanism dependant by default
on CONFIG_MULTITHREADING but leave the option for the dev to turn it
off.

Signed-off-by: Lucas Denefle <lucas.denefle@converge.io>

nordicjm commented on subsys/debug/coredump/Kconfig:
```diff
@@ -28,6 +28,14 @@ config DEBUG_COREDUMP_BACKEND_FLASH_PARTITION
 	  "coredump-partition".
 
+config DEBUG_COREDUMP_BACKEND_RETAINED_MEM
+	bool "Use flash partition for coredump"
```
Comment needs updating

nordicjm commented on subsys/debug/coredump/Kconfig:
```diff
@@ -28,6 +28,14 @@ config DEBUG_COREDUMP_BACKEND_FLASH_PARTITION
+	  Core dump is saved to retained memory to be used
+	  next boot.
+
+
```
Second newline can go

nordicjm commented on drivers/retained_mem/Kconfig:
```diff
@@ -15,6 +15,13 @@ config RETAINED_MEM_INIT_PRIORITY
 
+config RETAINED_MEM_CONCURRENT_ACCESS
+	bool "If the retained mem will have concurrent access"
+    default y if MULTITHREADING
```
1x tab for Kconfigs (help text is 1x tab followed 2x spaces)

nordicjm commented on subsys/debug/coredump/coredump_backend_retained_mem.c:
```diff
@@ -0,0 +1,190 @@
+#include <errno.h>
```
Copyright and license part needed

nordicjm reviewed (commented):
Nice, didn't envisage someone would have a use for retained memory right away, that is a cool feature! Had a quick look at the files and added some comments.

ldenefle committed:
debug: coredump: add retained_mem backend

Adds the option to use retained_mem subsystem as a backend for
coredumps. This might be sensitive to power losses but in the event of a
software error and a reboot due to a watchdog the coredump should be
available on next boot for more complex processing.

Signed-off-by: Lucas Denefle <lucas.denefle@converge.io>

ldenefle commented:
The `retained_mem` subsystem comes in handy as on my current project, we have a lot of custom logging formatting so can't really use the `logging` backend and the `flash_partition` one is a bit overblown for my need. 
I really like the simplicity of retained memory over the other backends so it's perfect for my use case :+1:

nordicjm commented on subsys/debug/coredump/Kconfig:
```diff
@@ -28,6 +28,13 @@ config DEBUG_COREDUMP_BACKEND_FLASH_PARTITION
 	  "coredump-partition".
 
+config DEBUG_COREDUMP_BACKEND_RETAINED_MEM
+	bool "Use retained mem subsys for coredump"
```
```suggestion
	bool "Use retained memory for coredump"
```
Retained memory is a driver, one important thing to note is that retained memory will take up a whole area, e.g. if you use a 10KB RAM area for it, it is the whole 10KB RAM area, you can't only use e.g. 2KB and then use 8KB for something else. The data retention subsystem builds on top of retained memory and does add this support, it also adds an optional checksum and optional magic header to the data area - it is currently in review at https://github.com/zephyrproject-rtos/zephyr/pull/55125

ldenefle replied:
any ideas on when #55125 should get merged ?

nordicjm replied:
Unfortunately not, it seems to be suffering from a lack of people that want to review it, feel free to review it though :)

ldenefle replied:
Didn't realise I could review it, I'm not a maintainer

nordicjm commented on drivers/retained_mem/Kconfig:
```diff
@@ -15,6 +15,13 @@ config RETAINED_MEM_INIT_PRIORITY
+	bool "If the retained mem will have concurrent access"
+	default y if MULTITHREADING
+	help
+	  Enables support for drivers that can retain their data whilst the
+	  device is powered (may be lost in low power states).
+
```
Comment needs updating

nordicjm commented on drivers/retained_mem/Kconfig:
```diff
@@ -15,6 +15,13 @@ config RETAINED_MEM_INIT_PRIORITY
 	  Retained memory devices initialization priority,
 
+config RETAINED_MEM_CONCURRENT_ACCESS
+	bool "If the retained mem will have concurrent access"
```
```suggestion
	bool "Retained memory locking (mutex) support"
```

ldenefle commented on subsys/debug/coredump/coredump_backend_retained_mem.c:
```diff
@@ -0,0 +1,197 @@
+
+struct retainem_mem_hdr_t {
+	/* 'C', 'D' */
+	char magic[2];
```
Sounds good, I could rebase my PR on top of the data retention branch

ldenefle commented on drivers/retained_mem/retained_mem_nrf_gpregret.c:
```diff
@@ -125,7 +125,7 @@ static const struct retained_mem_driver_api nrf_gpregret_api = {
+			      COND_CODE_1(CONFIG_RETAINED_MEM_CONCURRENT_ACCESS,		\
 					  (&nrf_gpregret_data_##inst), (NULL)	\
 			      ),						\
 			      &nrf_gpregret_config_##inst,			\
```
My clang-format acts a bit crazy around those big macros, do you have any advice on how to deal with them ?

nordicjm replied:
Is this using the default zephyr .clang-format file? @gmarull

ldenefle replied:
Yep as far as I know

ldenefle commented:
Just for clarity, this PR is currently being held up by #55125  once it's merged I'll rebase it and rewrite some bits to fit the retention subsys

nordicjm commented:
> Just for clarity, this PR is currently being held up by #55125 once it's merged I'll rebase it and rewrite some bits to fit the retention subsys

Can you review https://github.com/zephyrproject-rtos/zephyr/pull/55125 and leave comments if you think something should change or approve if you think it looks okay? Will help in getting it merged.

ldenefle commented:
Hello @nordicjm considering #55125 seem to be on hold, what would your recommendations be about this PR ?

nordicjm commented:
It's on hold because I'm awaiting @d3zd3z to merge https://github.com/mcu-tools/mcuboot/pull/1629 in mcuboot

ldenefle commented:
@nordicjm will do today

ldenefle commented:
I'll rebase that onto the now upstream data_retention subsystem

MaureenHelm commented:
@nashif please take a look

Laczen commented:
@ldenefle, a recent addition to the flash simulator allows to assign a memory region to the flash simulator (PR #57380). You can use the available `coredump_backend_flash_partition` without the need to add an extra backend.

ldenefle commented:
@Laczen that looks good, I think I can close this PR for the time being
